rio_tiler/reader.py

```diff
@@ -0,0 +1,552 @@
+        else:
+            mask = vrt.dataset_mask(
+                out_shape=(height, width),
+                window=out_window,
+                resampling=Resampling[resampling_method],
+            )
+
+        if force_binary_mask:
+            mask = numpy.where(mask != 0, numpy.uint8(255), numpy.uint8(0))
+
+        if unscale:
+            data = data.astype("float32", casting="unsafe")
+            numpy.multiply(data, vrt.scales[0], out=data, casting="unsafe")
+            numpy.add(data, vrt.offsets[0], out=data, casting="unsafe")
+
+    return data, mask
+
+
+def part(
+    src_dst: Union[DatasetReader, DatasetWriter, WarpedVRT],
+    bounds: Tuple[float, float, float, float],
+    height: int,
+    width: int,
+    padding: int = 0,
+    dst_crs: CRS = constants.WEB_MERCATOR_CRS,
+    bounds_crs: Optional[CRS] = None,
+    minimum_overlap: Optional[float] = None,
+    warp_vrt_option: Dict = {},
+    **kwargs: Any,
+) -> Tuple[numpy.ndarray, numpy.ndarray]:
+    """
+    Read part of an image.
+
+    Attributes
+    ----------
+        src_dst : rasterio.io.DatasetReader
+            rasterio.io.DatasetReader object
+        bounds : tuple
+            Output bounds (left, bottom, right, top) in target crs ("dst_crs").
+        height : int
+            Output height of the array.
+        width : int
+            Output width of the array.
+        padding : int, optional
+            Padding to apply to each edge of the tile when retrieving data
+            to assist in reducing resampling artefacts along edges.
+        dst_crs: CRS or str, optional
+            Target coordinate reference system, default is "epsg:3857".
+        bounds_crs: CRS or str, optional
+            Overwrite bounds coordinate reference system, default is equal
+            to the output CRS (dst_crs).
+        minimum_tile_cover: float, optional
+            Minimum % overlap for which to raise an error with dataset not
+            covering enought of the tile.
+        warp_vrt_option: dict, optional
+            These will be passed to the rasterio.warp.WarpedVRT class.
+        kwargs : Any, optional
+            Additional options to forward to reader._read()
+
+    Returns
+    -------
+        data : numpy ndarray
+        mask: numpy array
+
+    """
+    if not bounds_crs:
+        bounds_crs = dst_crs
+
+    bounds = transform_bounds(bounds_crs, dst_crs, *bounds, densify_pts=21)
+
+    src_bounds = transform_bounds(src_dst.crs, dst_crs, *src_dst.bounds, densify_pts=21)
+    x_overlap = max(0, min(src_bounds[2], bounds[2]) - max(src_bounds[0], bounds[0]))
+    y_overlap = max(0, min(src_bounds[3], bounds[3]) - max(src_bounds[1], bounds[1]))
+    cover_ratio = (x_overlap * y_overlap) / (
+        (bounds[2] - bounds[0]) * (bounds[3] - bounds[1])
+    )
+    if minimum_overlap and cover_ratio < minimum_overlap:
+        raise TileOutsideBounds(
+            "Dataset covers less than {:.0f}% of tile".format(cover_ratio * 100)
+        )
```

Comment:
`src_bounds`, `x_overlap`, `y_overlap`, and `cover_ratio` are only necessary in the case that `minimum_overlap` is not `None`. I found in my profiling that each call to `transform_bounds` here (inside current `_tile_read`) took 85ms. For 6 assets, thats .5 seconds of computation that aren't necessary when `minimum_overlap` is `None`.

Hence I'd suggest:
```py
if minimum_overlap:
    src_bounds = transform_bounds(src_dst.crs, dst_crs, *src_dst.bounds, densify_pts=21)
    x_overlap = max(0, min(src_bounds[2], bounds[2]) - max(src_bounds[0], bounds[0]))
    y_overlap = max(0, min(src_bounds[3], bounds[3]) - max(src_bounds[1], bounds[1]))
    cover_ratio = (x_overlap * y_overlap) / (
        (bounds[2] - bounds[0]) * (bounds[3] - bounds[1])
    )

    if cover_ratio < minimum_overlap:
        raise TileOutsideBounds(
            "Dataset covers less than {:.0f}% of tile".format(cover_ratio * 100)
        )
```

Reply:
nice find, totally make sense